internal/tools/tool/ads.go

```diff
@@ -5,63 +5,31 @@ package tool
 
 import (
 	"github.com/microsoft/go-sqlcmd/internal/io/file"
-	"os"
-	"path/filepath"
+	"github.com/microsoft/go-sqlcmd/internal/test"
 )
 
 type AzureDataStudio struct {
-	Base
+	tool
 }
 
 func (t *AzureDataStudio) Init() {
-	t.Base.SetName("ads")
-	userProfile := os.Getenv("USERPROFILE")
-	programFiles := os.Getenv("ProgramFiles")
+	t.tool.SetToolDescription(Description{
+		Name:        "ads",
+		Purpose:     "Azure Data Studio provides a user interface for working with SQL Server and Azure SQL Database.",
+		InstallText: t.installText()})
 
-	// Search in this order
-	//   User Insiders Install
-	//   System Insiders Install
-	//   User non-Insiders install
-	//   System non-Insiders install
-	searchLocations := []string{
-		filepath.Join(userProfile, "AppData\\Local\\Programs\\Azure Data Studio - Insiders\\azuredatastudio-insiders.exe"),
-		filepath.Join(programFiles, "Azure Data Studio - Insiders\\azuredatastudio-insiders.exe"),
-		filepath.Join(userProfile, "AppData\\Local\\Programs\\Azure Data Studio\\azuredatastudio.exe"),
-		filepath.Join(programFiles, "Azure Data Studio\\azuredatastudio.exe"),
-	}
-
-	t.Base.SetExeName("azuredatastudio")
-	for _, location := range searchLocations {
+	for _, location := range t.searchLocations() {
 		if file.Exists(location) {
-			t.Base.SetExeName(location)
+			t.tool.SetExePathAndName(location)
 			break
 		}
 	}
-
-	t.Base.SetToolDescription(Description{
-		t.Name(),
-		"Azure Data Studio provides a user interface for working with SQL Server and Azure SQL Database.",
-		InstallText{
-			Windows: `Download the latest 'User Installer' .msi from:
-
-    https://go.microsoft.com/fwlink/?linkid=2150927
-
-More information can be found here:
-
-    https://docs.microsoft.com/en-us/sql/azure-data-studio/download-azure-data-studio#get-azure-data-studio-for-windows`,
-			Linux: `Follow the instructions here:
-
-   https://docs.microsoft.com/en-us/sql/azure-data-studio/download-azure-data-studio?#get-azure-data-studio-for-linux`,
-			Mac: `Download the latest .zip from:
-
-    https://go.microsoft.com/fwlink/?linkid=2151311
-
-More information can be found here:
-
-    https://docs.microsoft.com/en-us/sql/azure-data-studio/download-azure-data-studio?#get-azure-data-studio-for-macos`,
-		}})
 }
 
 func (t *AzureDataStudio) Run(args []string) (int, error) {
-	return t.Base.Run(args)
+	if !test.IsRunningInTestExecutor() {
+		return t.tool.Run(args)
```

Comment:
this is another opportunity to introduce dependency inversion where the tests can inject an alternate Run method for a tool. That will enable the tests to validate the correct args are passed based on the current config.
It also allows removal of test-environment-specific code from the product code.

Reply:
Agree, this is a good idea and great feedback (I'd become a bit stuck on approach here).  I've entered an issue to refactor this package.  We've been asked to add SSMS support as well, so when I do that, I'll also refactor with the appropriate interfaces to enable DI here.

Reply:
[https://github.com/microsoft/go-sqlcmd/issues/266](https://github.com/microsoft/go-sqlcmd/issues/266)